### ISSUE TITLE
[MEM][S2] the n==1 sampler still uses the pageable D2H plus sync the batched path replaced

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1654]
## Fact

- The single-sequence sampler exits through a 4-byte D2H into a **pageable stack variable** followed by `cudaStreamSynchronize`, at four sites — the exact pattern the batched path measured at ~850 us of blocked host time per sequence per step and replaced with a pinned gather (`src/compute/sampling.cu:181-182`, `:203-204`, `src/compute/sampling_topk_topp.cu:652-653`, `:675-676` vs `src/exec/executor.cu:322-330`).

## Evidence

```
src/compute/sampling.cu:200-207
    argmax_reduce_kernel<<<1, WARP_SIZE, 0, stream>>>(partial_vals, partial_idxs, ARGMAX_NBLOCKS, d_result);
    IMP_CUDA_CHECK_LAUNCH();

    int32_t h_result = 0;
    IMP_CUDA_CHECK_LOG(cudaMemcpyAsync(&h_result, d_result, sizeof(int32_t), cudaMemcpyDeviceToHost, stream));
    cudaStreamSynchronize(stream);

    return h_result;
```

`h_result` is an automatic variable, i.e. pageable host memory. Same shape at `sampling.cu:178-183` (the no-`d_result` overload) and at `sampling_topk_topp.cu:650-654` and `:673-677`.

The project already measured this exact pattern and fixed it on the other path:

```
src/exec/executor.cu:322-330
        // Two-pass batched sampling: pass 1 ENQUEUES every row's filter+
        // sampler chain into its own scratch slot ...; pass 2
        // gathers all tokens with ONE pinned D2H + ONE stream sync. The
        // previous per-row synchronous readback blocked the engine thread
        // ~850 us per sequence per step (pageable 4-byte D2H + sync each) —
        // 29% GPU idle at n=16 sustained serving (nsys, 2026-07-12).
```

and the pinned buffer that fix introduced (`h_sample_pinned_`) is never used by the single-sequence path:

```
src/exec/executor.cu:497-509   (sample_single_from_logits, the n==1 exit)
    return (state.temperature <= 0.0f || state.top_k == 1)
               ? (d_sample_result_ ? sample_greedy(flat, d_sample_result_, stream)
                                   : sample_greedy(flat, stream))
               : (d_sample_result_
                      ? sample_topk_topp(flat, ..., d_sample_result_, stream)
                      : sample_topk_topp(flat, ..., stream));
```

Frequency chain — the sites fire once per decode step per request whenever the async conditional loop is refused:

- `Engine::step_decode_forward` (`src/runtime/engine_scheduler.cpp:1780`) -> `sample_per_request` -> `n == 1` branch (`:1854-1860`) -> `GraphExecutor::sample_single_from_logits` (`src/exec/executor.cu:479`) -> the four sites.
- The async conditional loop, which does sample in-graph, is refused for a request that is constrained, wants logprobs, or uses logit_bias / mirostat / DRY / min_p / typical_p, or has MTP on (`src/runtime/engine_scheduler.cpp:2374-2412`), and outright for `req->constraints || req->json_mode || !req->json_schema.empty() || !req->regex_pattern.empty() || !req->grammar.empty() || !req->tool_constraint_tools.empty()` (`src/runtime/engine_graph_decode.cpp:231-233`).
- Per `docs/LIMITATIONS.md`, that set is "every tool call and every structured output", i.e. the agentic workload `docs/GOAL.md` targets.

Measured for scale on a config that does NOT take this path (batch 1, unconstrained, async loop live): only 2 of 514 decode steps ran outside the conditional graph — `_audit/raw/nsys_decode.txt` kernel summary reconciles exactly: `gemv_nvfp4_gate_up_fused_mr` 18504 / 36 layers = 514 decode forwards, `post_decode_step_kernel` (in-graph only) 512, `paged_attention_decode_fp8_kernel` 72 / 36 = the 2 remaining, and `cudaStreamSynchronize` totals 205 calls over the whole process, i.e. **zero per-token stream syncs on the default unconstrained path**. The pattern is therefore latent at defaults and live on the constrained path.

## Impact

- Per decode step on any request that cannot use the conditional loop: one 4-byte pageable D2H plus one full `cudaStreamSynchronize`. The measured floor on this host is a single D2H plus sync at **45.4 us**; the batched path's own comment prices the same pattern at ~850 us of blocked host time per sequence per step under load.
- Against the 3.33 ms of GPU kernel time per decode step measured in `_audit/raw/nsys_decode.txt` (1711.6 ms of decode-attributable kernel time over 514 decode forwards), 45.4 us is 1.4 % of a step; the 850 us figure the batched path measured is 27 %. Which of the two applies to the constrained path is not measured — see Unverified.
- The pageable half is the avoidable part: the driver stages a pageable D2H through an internal pinned buffer, and `GraphExecutor` already owns a pinned 4-byte-per-slot buffer (`h_sample_pinned_`) that the single-sequence path does not touch.
- Blast radius: constrained decode (json_mode, json_schema, regex, GBNF, enforced tool calls), logprobs requests, and every request using logit_bias / mirostat / DRY / min_p / typical_p. Also the first decode step of every request, before the loop is armed.

## Repro

Static, runnable now:

```
$ grep -n "cudaStreamSynchronize" src/compute/sampling.cu src/compute/sampling_topk_topp.cu
$ grep -n "h_sample_pinned_" src/exec/executor.cu
```

MEASUREMENT REQUEST (NOT RUN) — measure the constrained path, which the existing profile does not cover:

```
nsys profile --cuda-graph-trace=node --sample=none --cpuctxsw=none --backtrace=none \
  -o constrained_decode -- \
  imp-server --model /models/Qwen3-8B-Q8_0.gguf --set speculative.ngram=false
# drive one streaming request with response_format={"type":"json_schema", ...},
# >= 256 output tokens, then read cuda_api_sum:
#   cudaStreamSynchronize count should be ~= output token count if this path is hot
#   cudaMemcpy / cudaMemcpyAsync D2H count likewise
# Report tok/s for the same prompt with and without the schema as the A/B.
```

## Fix direction

- Route the `n == 1` exit through the same pinned slot the batched path uses: `sample_greedy_async` / `sample_topk_topp_async` into slot 0, then `collect_sampled_tokens(1, stream)` (`src/exec/executor.cu:591`). One pinned D2H, one sync, no new machinery.
- If the synchronous overloads must stay for the test-only callers, give them an engine-owned pinned staging buffer rather than a stack `int32_t`.
- The constrained pipeline (`Engine::step_constrained_pipeline`, `src/runtime/engine_graph_decode.cpp:507`) already exists to hide this latency for json/schema requests; check whether it is reached for regex/GBNF/tool-constraint requests too, or whether those fall back to the eager path.

## Acceptance criteria

1. `grep -n "int32_t h_result" src/compute/sampling.cu src/compute/sampling_topk_topp.cu` returns no hit that is the destination of a `cudaMemcpyAsync`.
2. An nsys `cuda_api_sum` over a 256-token json_schema-constrained generation shows `cudaStreamSynchronize` count below 0.5 x the output token count.
3. `make verify-fast` decode tg128 within 1 % of the 287.19 pin (the default path must not move).

## Unverified

- The cost on the constrained path is not measured. The 45.4 us floor and the 850 us figure are from two different regimes (idle stream, and n=16 batched serving respectively); neither is a measurement of a batch-1 constrained decode step.
- Whether the constrained pipeline covers regex/GBNF/tool-constraint requests, or only json_mode/json_schema, is not established here.

**Comment by kekzl** [comment]
Measured the half this issue lists as unverified: **the pageable/pinned difference is not where the cost is.**

Isolated 4-byte D2H + `cudaStreamSynchronize` on this box (RTX 5090, sm_120a), 2000 iterations after 200 warmup, the same shapes the sampler uses - a stack automatic against a `cudaHostAlloc` buffer:

```
pageable (stack)  median  32.21 us   min  9.67   max  154.83
pinned            median  31.35 us   min  9.19   max 8378.03
delta on medians  -0.86 us (-2.7%)
```

So swapping the destination buffer buys **0.86 us of the ~32 us**. The cost is the synchronize, not the staging: the driver's internal pinned bounce for a 4-byte transfer is not measurable against a full stream sync.

That changes the fix direction. Pointing the four sites at `h_sample_pinned_` would be a real change that measures as noise; removing the sync - the batched path's actual lever, which gathers N results with ONE sync instead of N - is where the 850 us per sequence per step came from.

Caveat on scope: this is an isolated D2H, not the sampler in context. It answers "is the pageable buffer the expensive half" (no) and not "what does the whole pattern cost inside a decode step", which the issue already prices at 45.4 us from nsys.

Leaving the issue open, because the sync half is still real - just not the half the fix direction names.

**Comment by kekzl** [comment]
Repeated on a verified-quiet card, because the first run may have overlapped a neighbouring GPU job (that is what the 8378 us outlier looked like):

```
run 2, 06:15:47Z-06:15:50Z
  pageable (stack)  median  42.67 us   min 13.32   max 3820.35
  pinned            median  42.04 us   min 11.62   max 3301.28
  delta on medians  -0.63 us (-1.5%)
```

The absolute numbers move with host load (42.7 against 32.2 us for the same code), which is why the delta is the measurement and not the level. Both runs agree: **pinned buys under 1 us of a ~32-43 us pattern.**
